src/viam/sdk/common/exception.hpp

```diff
@@ -0,0 +1,84 @@
+/// @file common/exception.hpp
```

Comment:
The classes in this file are mostly a subset of the [error types from the Python SDK](https://github.com/viamrobotics/viam-python-sdk/blob/main/src/viam/errors.py#L1). I added both `DuplicateRegistrationException` _and_ `DuplicateResourceException` because they seemed like different cases to me.